Feature/doxygen

simonnagl opened the pull request:
Added doxygen-style documentation to `proc_self_mountinfo.h`
This should not be merged as is.

Things to be done before we can merge this:
- [x] Include generation of the doc to into the build process.
- [x] Document at least all header files.

To view the documentation these steps are needed:
1. Change dir to project root.
2. `doxygen doxygen.conf`
3. View `doc/html/index.html` to see the documentation of the file.

simonnagl commented:
Commented appconfig.h. @ktsaou Please tell me if you like this kind of documentation. If you like it I would document all the header files like this. I think this makes a enormous step in code quality and can help new contributors.

If you like it also comment if I made some mistakes and comment the lines I requested your help. Also tell me if you think it is to much work for the benefit.

ktsaou commented:
@simonnagl nice work.

> Include generation of the doc to into the build process.

Keep in mind I would like this to be opt-in, not opt-out.

simonnagl commented:
> I have added comments to commit beb13dc

I have read them and updated the comments.
> Keep in mind I would like this to be opt-in, not opt-out.

This was my intention too.

ktsaou commented:
perfect!

simonnagl commented:
Documented avl.h. @ktsaou There I also need your help. Do you want me to always notify you if there is something new your help is requested or do you want to get it bundled after I finished all header files.

Now `doxygen` is included into the build chain. The Output format of doxygen can be configured. For example `--disable-doxygen-pdf` for a complete list run `./configure --help`

There are some new make targets:
- doxygen-doc: Generate all doxygen documentation.
- doxygen-run: Run doxygen, which will generate some of the documentation (HTML, CHM, CHI, MAN, RTF, XML) but will not do the post processing required for the rest of it (PS, PDF).
- doxygen-ps:  Generate doxygen PostScript documentation.
- doxygen-pdf: Generate doxygen PDF documentation.

Now there is one question to discuss. Do we want to add an option `--generate-doxygen` to the netdata installer? Where should we document this feature?

ktsaou commented on .gitignore:
```diff
@@ -83,6 +83,7 @@ CMakeFiles/
 cmake_install.cmake
 
 .jetbrains*
+<<<<<<< HEAD
```
oops!

ktsaou commented:
sorry.. I broke this...

simonnagl commented:
No problem. Rebased to master. Your doing real good work!

simonnagl commented:
@ktsaou Added some new TODO's where I need your help...

ktsaou commented on src/avl.h:
```diff
@@ -22,64 +42,122 @@
 
-
+/**
+ * \todo kstou Your help needed.
```
walk through all entries in the AVL tree. The function `int callback(entry, data)` will be called for each node (`data` is the `data` parameter to the traverse function).

The return value of `callback()` is interpreted as follows:

- zero, or positive = add it to `total`.
- negative = stop traversing the tree.

If no call to callback returns a negative number, the return value of the traverse function is the `total`. 
If any call to callback returns a negative number, the return value is the number returned by that callback function.

I have never verified, if this calls the callback function in a sorted order (I hope it does, but I never really cared since netdata uses these AVL trees with string hashes - so sorting them based on the hashes is meaningless).

ktsaou commented on src/clocks.h:
```diff
@@ -1,92 +1,173 @@
+ * 
+ * @see man 2 gettimeofday
+ *
+ * @param clk_id Not used. \todo Should we change API here?
```
this is work in progress. There is PR #1368 with more changes.

ktsaou commented on src/main.h:
```diff
@@ -1,6 +1,34 @@
+ *
+ * \section sec_outline Outline
+ *
+ * \todo 
```
?

simonnagl replied:
This comment block starts with \mainpage.

This will be the initial start page of `doxygen`. I am not shure anymore if this is the right place to put it because your question mark. The todo is for me.

I think on the homepage of the `doxygen` documentation we should give an overview where to find wich part of net data. Also there should be additional information about coding rules etc. (Things the user of `netdata` does not need to no but the contributor)

What do you think?

ktsaou commented on src/common.h:
```diff
@@ -236,37 +347,176 @@ extern void *mallocz_int(const char *file, const char *function, const unsigned
 #endif
 
+/**
+ * \todo ktsaou: your help needed. Especially when this should be used.
```
When a string is to be added to a JSON object, we need to escape it (i.e. escape double quotes and backslashes). It is like `strncpy()`, but it escapes the string while copying it.

ktsaou commented on src/common.h:
```diff
@@ -236,37 +347,176 @@ extern void *mallocz_int(const char *file, const char *function, const unsigned
 extern void json_escape_string(char *dst, const char *src, size_t size);
 
+/**
+ * \todo ktsaou: your help needed. Especially what is the difference to mmap 
```
1. it creates the file if does not exist (to the correct size).
2. it truncates the file if it already exists and is bigger than expected.
3. sets various memory management heuristics we know netdata memory databases use (like `MADV_SEQUENTIAL`)
4. prevents this memory from being copied when netdata forks to execute its plugins.
5. enables KSM for it.
6. in memory mode `ram`, it loads the file into memory (in all other modes, loading it is on-demand by the kernel).

ktsaou commented on src/common.h:
```diff
@@ -277,6 +527,40 @@ extern void get_system_HZ(void);
+ */
 extern int read_single_number_file(const char *filename, unsigned long long *result);
 
+/// \todo ktsaou: your help needed
```
haha... this is old... you need to fetch the latest.

To understand simple patterns, check this https://github.com/firehol/netdata/issues/1571#issuecomment-272705918

You can run `netdata -W simple-pattern` for more help or to play with it...

ktsaou replied:
The general rule is:

1. The `create` call builds a linked list (more like a tree) of substrings to be matched, each matched as PREFIX, SUFFIX, SUBSTRING or EXACT.
2. The `match` call checks a string against a pattern that has been parsed with `create`.
3. The `free` call frees the memory allocated by `create`.

The whole mechanism supports both positive and negative matches.

simonnagl replied:
> haha... this is old... you need to fetch the latest

This is in the current master... Did I understand you wrong?

ktsaou commented on src/common.h:
```diff
@@ -277,6 +527,40 @@ extern void get_system_HZ(void);
+    NETDATA_SIMPLE_PATTERN_MODE_SUFFIX,   ///< \todo ktsaou: your help needed
+    NETDATA_SIMPLE_PATTERN_MODE_SUBSTRING ///< \todo ktsaou: your help needed
+} NETDATA_SIMPLE_PREFIX_MODE;
+typedef void NETDATA_SIMPLE_PATTERN; ///< \todo ktsaou: you help needed
```
This is a way to prevent exposing the internals of simple patterns to the rest of netdata. So, all the calls accept `SIMPLE_PATTERN *`, which is actually a `void *`.

ktsaou commented:
I think I answered all. If you find any missing, please let me know.
You need to update your repo...

ktsaou commented:
ok. sounds nice.

simonnagl commented:
I went on and documented all structs. 

@ktsaou There are some TODOs I nee your help.

ktsaou commented on src/global_statistics.h:
```diff
@@ -2,18 +2,17 @@
+    volatile uint16_t connected_clients; ///< number of connected clients
+
+    volatile uint64_t web_requests;            ///< Number of web requests.
+    volatile uint64_t web_usec;                ///< ktsaou: Your help needed
```
sum of the duration to serve `web_requests` (from the reception of the request, to the dispatch of the last byte).

ktsaou commented on src/global_statistics.h:
```diff
@@ -2,18 +2,17 @@
+
+    volatile uint64_t web_requests;            ///< Number of web requests.
+    volatile uint64_t web_usec;                ///< ktsaou: Your help needed
+    volatile uint64_t web_usec_max;            ///< ktsaou: Your help needed
```
the max time to serve a request. This is reset to zero every time the chart is updated, so it shows the max time for each iteration.

ktsaou commented on src/eval.h:
```diff
@@ -3,15 +3,17 @@
+    struct eval_variable *next; ///< Next item in the list
 } EVAL_VARIABLE;
 
+/// ktsaou: Your help needed
```
This is the internal representation infix expressions used in alarms. netdata splits the expression to a tree, where each node is performing a calculation.

ktsaou commented on src/eval.h:
```diff
@@ -3,15 +3,17 @@
+/// ktsaou: Your help needed
 typedef struct eval_expression {
-    const char *source;
+    const char *source;q
```
the expression as given by the user

ktsaou commented on src/eval.h:
```diff
@@ -3,15 +3,17 @@
 typedef struct eval_expression {
-    const char *source;
+    const char *source;q
     const char *parsed_as;
```
the expression, reformatted after parsing it (and generating the expression tree) and re-producing from the tree.

ktsaou commented on src/eval.h:
```diff
@@ -3,15 +3,17 @@
+    const char *source;q
     const char *parsed_as;
 
     int *status;
```
the result status of the expression

ktsaou commented on src/health.h:
```diff
@@ -13,89 +13,92 @@ extern int rrdvar_compare(void *a, void *b);
+    void *value; ///< the value
 
-    uint32_t options;
+    uint32_t options; ///< ktsaou: your help needed
```
not used currently

ktsaou commented on src/health.h:
```diff
@@ -13,89 +13,92 @@ extern int rrdvar_compare(void *a, void *b);
+    void *value; ///< variable value
 
-    uint32_t options;
+    uint32_t options; ///< ktsaou: Your help needed
```
not used currently

ktsaou commented on src/health.h:
```diff
@@ -127,149 +130,153 @@ typedef struct rrddimvar {
 
-    char *exec;
-    char *recipient;
+    char *exec;      ///< ktsaou: your help needed
```
the exec line of the alarm template (a script)

ktsaou commented on src/health.h:
```diff
@@ -127,149 +130,153 @@ typedef struct rrddimvar {
-    char *exec;
-    char *recipient;
+    char *exec;      ///< ktsaou: your help needed
+    char *recipient; ///< ktsaou: your help needed
```
the `to` line of the alarm template

ktsaou commented on src/health.h:
```diff
@@ -127,149 +130,153 @@ typedef struct rrddimvar {
 
-    char *context;
-    uint32_t hash_context;
+    char *context;         ///< ktsaou: your help needed
```
the `on` line of the alarm template

ktsaou commented on src/health.h:
```diff
@@ -127,149 +130,153 @@ typedef struct rrddimvar {
 
-    char *family_match;
-    SIMPLE_PATTERN *family_pattern;
+    char *family_match;             ///< ktsaou: your help needed
```
the `families:` line of the alarm template (a simple-pattern)

ktsaou commented on src/health.h:
```diff
@@ -127,149 +130,153 @@ typedef struct rrddimvar {
-    char *family_match;
-    SIMPLE_PATTERN *family_pattern;
+    char *family_match;             ///< ktsaou: your help needed
+    SIMPLE_PATTERN *family_pattern; ///< ktsaou: your help needed
```
the parsed pattern

ktsaou replied:
this is used to match the charts the alarm template should by applied to. Example `veth*` to match all `veth` interfaces.

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
-    char *recipient;
-    time_t exec_run_timestamp;
-    int exec_code;
+    char *exec;                ///< ktsaou: your help needed
```
the script to run when the alarm switches state.

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
-    time_t exec_run_timestamp;
-    int exec_code;
+    char *exec;                ///< ktsaou: your help needed
+    char *recipient;           ///< ktsaou: your help needed
```
the first parameter of the script to run when the alarm switches state.

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
-    char *source;
-    char *units;
-    char *info;
+    char *source; ///< ktsaou: your help needed
```
the source description of the alarm that generated this alarm (i.e. `LINE@FILE`).

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
 
-    int delay;
-    time_t delay_up_to_timestamp;
+    int delay;                    ///< ktsaou: your help needed
```
the "delay to run the script" as was calculated when the alarm entry was created.

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
-    int delay;
-    time_t delay_up_to_timestamp;
+    int delay;                    ///< ktsaou: your help needed
+    time_t delay_up_to_timestamp; ///< ktsaou: your help needed
```
the "delay to run the script" expressed as an absolute timestamp.

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
 
-    uint32_t updated_by_id;
-    uint32_t updates_id;
+    uint32_t updated_by_id; ///< ktsaou: your help needed
```
the id of the alarm entry that updated this alarm entry (so that a new alarm entry cancel any previous for the same alarm).

ktsaou commented on src/health.h:
```diff
@@ -280,55 +287,57 @@ typedef struct rrdcalctemplate {
-    uint32_t updated_by_id;
-    uint32_t updates_id;
+    uint32_t updated_by_id; ///< ktsaou: your help needed
+    uint32_t updates_id;    ///< ktsaou: deyour help needed
```
the id of the alarm entry that has been updated by this alarm entry (the one that was cancelled because of this alarm entry).

ktsaou commented on src/registry_internals.h:
```diff
@@ -11,54 +11,56 @@
+    unsigned long long save_registry_every_entries; ///< Number after how many new entries registry should be saved
+    char *registry_domain;                          ///< domain of the registry
+    char *hostname;                                 ///< hostname of the registry
+    char *registry_to_announce;                     ///< ktsaou: Your help is needed
```
well, the registry to announce for fetching the entries in `my-netdata` menu.

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
+    Bytef zbuffer[ZLIB_CHUNK];      ///< temporary buffer for storing compressed output
+    size_t zsent;                   ///< the compressed bytes we have sent to the client
+    size_t zhave;                   ///< the compressed bytes that we have received from zlib
+    int zinitialized:1;             ///< ktsaour: Your help needed
```
netdata supports delayed initialization of the compression library for each request. This keeps track if the compression library has been initialized or not.
to understand it better, focus on the single threaded web server: everything is async, so we need to keep states in variables.

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
 
-    int ifd;
-    int ofd;
+    int ifd; ///< ktsaou: Your help needed
```
input file (think: serving static web files).

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
-    int ifd;
-    int ofd;
+    int ifd; ///< ktsaou: Your help needed
+    int ofd; ///< ktsaou: Your help needed
```
output file (the web client socket)

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
+    char last_url[URL_MAX+1];       ///< we keep a copy of the decoded URL here
 
-    struct timeval tv_in, tv_ready;
+    struct timeval tv_in;    ///< ktsaou: Your help needed
```
do you remember global statistics web_usec?
this is the start time.

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
 
-    struct timeval tv_in, tv_ready;
+    struct timeval tv_in;    ///< ktsaou: Your help needed
+    struct timeval tv_ready; ///< timestamp client became ready
```
this is the time we finished processing the request (i.e. "response ready").

ktsaou replied:
this is only used for access.log - the web_usec global statistics is reporting time to the dispatch of the last byte.

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
-    char cookie1[COOKIE_MAX+1];
-    char cookie2[COOKIE_MAX+1];
-    char origin[ORIGIN_MAX+1];
+    char cookie1[COOKIE_MAX+1]; ///< ktsaou: Your help needed
```
the host cookie used by the registry to track web browsers

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
-    char cookie2[COOKIE_MAX+1];
-    char origin[ORIGIN_MAX+1];
+    char cookie1[COOKIE_MAX+1]; ///< ktsaou: Your help needed
+    char cookie2[COOKIE_MAX+1]; ///< ktsaou: Your help needed
```
the domain cookie used by the registry to track web browsers

ktsaou commented on src/web_client.h:
```diff
@@ -18,73 +18,76 @@ extern int web_enable_gzip, web_gzip_level, web_gzip_strategy, web_donotrack_com
-    char origin[ORIGIN_MAX+1];
+    char cookie1[COOKIE_MAX+1]; ///< ktsaou: Your help needed
+    char cookie2[COOKIE_MAX+1]; ///< ktsaou: Your help needed
+    char origin[ORIGIN_MAX+1];  ///< ktsaou: Your help needed
```
CORS.
This is used to store the `origin` HTTP header, so that netdata enables CORS for this specific origin instead of `*`.

ktsaou commented:
done

simonnagl commented on src/proc_self_mountinfo.h:
```diff
@@ -1,55 +1,106 @@
+    char *mount_options; ///< per-mount options.
 
-    int optional_fields_count;
+    int optional_fields_count; ///< count optional fields \todo deprecated?
```
@ktsaou Can you answer this question?

ktsaou replied:
`/proc/mountinfo` may have a number of optional fields, that are terminated with a field with contents ` - `. This counts how many of such fields are there. I have never found a system that this is non-zero.

ktsaou commented on src/eval.h:
```diff
@@ -39,43 +44,69 @@ typedef struct eval_expression {
 
 // parsing and evaluation
-#define EVAL_ERROR_OK                             0
+#define EVAL_ERROR_OK                             0 ///< Minor error (ktsaou: is this true?)
```
ok, this is a stupid label. This means: OK, NO ERROR.

simonnagl commented:
@ktsaou We may use this https://codedocs.xyz

ktsaou commented:
> We may use this https://codedocs.xyz

looks good! We can enable this on the main repo when this is merged.

simonnagl commented:
Today I restructured doxygen a bit.

Doxygen related files go to `doxygen/`. This is because we may document code from other directories sometimes. The main page of the doxygen documentation is now there.
The overview over the code in `src/` goes to `src/README.md` and is included into doxygen.

If you do not agree with this decisions please let me know.

ktsaou commented on src/dictionary.h:
```diff
@@ -27,22 +38,86 @@ typedef struct dictionary {
-#define DICTIONARY_FLAG_NAME_LINK_DONT_CLONE    0x00000004
-#define DICTIONARY_FLAG_WITH_STATISTICS         0x00000008
+#define DICTIONARY_FLAG_DEFAULT                 0x00000000 ///< no specific meaning.
+#define DICTIONARY_FLAG_SINGLE_THREADED         0x00000001 ///< If enabled synchronize access to dictionary.
```
if set, do not use the avl *_lock() functions, but the plain ones. This can lead to crashes if multiple threads are adding, removing and reading the tree concurrently.

simonnagl commented:
@ktsaou Should there be a file `sys_fs_cgroup.h` what is the reason for defining `extern void *cgroups_main(void *ptr);` in main.c?

simonnagl commented:
Documented most header files of threads. @ktsaou there are some todo's left for you.

ktsaou commented on src/global_statistics.h:
```diff
@@ -1,36 +1,86 @@
+    volatile uint64_t compressed_content_size; ///< Size of compressed content.
 };
 
+/// The global statistic representation. \todo ktsaou: Your help needed. Do we want to extern this? Why?
```
it is a left over I think

ktsaou commented on src/global_statistics.h:
```diff
@@ -1,36 +1,86 @@
+/**
+ * Lock the global statistics.
+ *
+ * \todo ktsaou: Your help needed. Do we want to extern this? Why?
```
left over...

ktsaou commented on src/global_statistics.h:
```diff
@@ -1,36 +1,86 @@
+/**
+ * Unlock the global statistics.
+ *
+ * \todo ktsaou: Your help needed. Do we want to extern this? Why?
```
left over...

ktsaou commented on src/global_statistics.h:
```diff
@@ -1,36 +1,86 @@
 extern void web_client_disconnected(void);
 
+/// ktsaou: Your help needed
 #define GLOBAL_STATS_RESET_WEB_USEC_MAX 0x01
```
it is used for reset the max response time. So, the caller will set this to indicate that the max value has been used and it can now be reset. Without this, there was no way to find the max duration per second.

ktsaou commented on src/plugin_proc.h:
```diff
@@ -1,31 +1,235 @@
+ *
+ * @return numa_node_count
+ */
 extern int get_numa_node_count(void);
```
@tycho 

I think this is used to find the number of numa nodes in this system.

tycho replied:
Yep. It's pretty self-descriptive.

ktsaou commented on src/procfile.h:
```diff
@@ -27,86 +28,164 @@
+ */
 extern void procfile_set_quotes(procfile *ff, const char *quotes);
+/**
+ * ktsaou: Your help needed
```
procfile supports quotes and opening / closing characters.
So, if you set this to `(` and `)`, the tokenizer will assume everything in parenthesis as a single keyword.

ktsaou commented on src/web_server.h:
```diff
@@ -1,36 +1,78 @@
 
 #ifndef HAVE_ACCEPT4
+/**
+ * ktsaou: Your help needed.
```
this is a replacement for `accept4()` for systems that do not have one. `man accept4`.

ktsaou commented on src/web_server.h:
```diff
@@ -1,36 +1,78 @@
 
 #ifndef SOCK_CLOEXEC
-#define SOCK_CLOEXEC 02000000
+#define SOCK_CLOEXEC 02000000 ///< ktsaou: Your help needed
```
`man accept4`

ktsaou commented:
done

ktsaou commented on src/inlined.h:
```diff
@@ -44,6 +51,9 @@ static inline uint32_t simple_uhash(const char *name) {
 }
 
+/**
+ * ktsaou: Your help needed
```
calculate a simple hash, but at the same time compare the string to be hashed with another.

So, normally one would do:

```c
hash = simple_hash(string);
if(hash == myhash && strcmp(string, "mystring") == 0) ...
```

The above requires 2 loops on `string`.

So, `if(simple_hash_strcmp(string, "mystring", &hash) == 0) ...` merges these 2 loops into one.

Currently it is only used for the `SET` line in pluginsd.c (the most common line send by external plugins to netdata).

ktsaou replied:
the result is about 40% less CPU time for the `SET` line.

ktsaou commented on src/web_client.h:
```diff
@@ -1,22 +1,30 @@
-#define WEB_CLIENT_MODE_FILECOPY    1
-#define WEB_CLIENT_MODE_OPTIONS     2
+#define WEB_CLIENT_MODE_NORMAL      0 ///< Standard web client mode.
+#define WEB_CLIENT_MODE_FILECOPY    1 ///< ktsaou: Your help needed.
```
deliver a static file to the web client

ktsaou commented on src/web_client.h:
```diff
@@ -1,22 +1,30 @@
-#define WEB_CLIENT_MODE_OPTIONS     2
+#define WEB_CLIENT_MODE_NORMAL      0 ///< Standard web client mode.
+#define WEB_CLIENT_MODE_FILECOPY    1 ///< ktsaou: Your help needed.
+#define WEB_CLIENT_MODE_OPTIONS     2 ///< ktsaou: Your help needed.
```
HTTP `OPTIONS` request, i.e. don't do the job, just respond if it is allowed. Used to enable cross-domain CORS.

ktsaou commented on src/web_client.h:
```diff
@@ -1,22 +1,30 @@
+extern int web_enable_gzip;      ///< enable gzip compressed web traffic
+extern int web_gzip_level;       ///< gzip level to use
+extern int web_gzip_strategy;    ///< gzip strategy to use
+extern int web_donotrack_comply; ///< boolean: Use do not track.
```
This enables HTTP DNT standard's compliance for netdata: https://github.com/firehol/netdata/pull/483

ktsaou commented:
@simonnagl I am lost. Do you have pendings for me?

simonnagl commented:
Currently I do not now. There I some files left I need to document. I will hopefully finish this the end of the next week. After that I have to review everything, fix typos, rebase to make the history a bit more clean...

Let's to it that way. When reviewing I will make a comment here to the lines I need your help. I think this request will need some more weeks. I do not want to merge bad documentation. I hope this is OK.

ktsaou commented:
it is perfect! you are doing great!

simonnagl commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
+ *
+ * @param hb Heartbeat to reset.
+ */
 extern void heartbeat_init(heartbeat_t *hb);
```
Maybe we should change this to
```c
#define HEARTBEAT_T_DEFAULT 0LL
```

rlefevre replied:
Again, this is an implementation detail from my point of view. I don't think this needs to be exposed in the header.

rlefevre commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
- *  which may have discontinuities if the time is changed using settimeofday(2).
- *  If not available on the system, this clock falls back to MONOTONIC clock.
+ * @param tv struct timeval gets updated
+ * @return 0 on succes, -1 else with errno set appropriately.
```
`succes` -> `success`

rlefevre commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
 extern usec_t dt_usec(struct timeval *now, struct timeval *old);
 
+/**
+ * Reset heartbeat to 0LL.
```
heartbeat_t is used as an opaque type. It could very well become a struct if needed (for example to store the offset with realtime clock in order to enhance alignment as suggested by @ktsaou). Therefore I don't think we should describe here a specific value (this can be done in the function definition if needed).

rlefevre commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
+/**
+ * Reset heartbeat to 0LL.
+ *
+ * \todo Maybe we should change the name (because init should have a corresponding free) to reset or even switch to use a define.
```
`_init()` usually implies initializing an already allocated type in C/C++ and in netdata, cf. for example `avl_init()`, `rrdr_line_init()` or `web_buffer()`. When allocation is performed, it seems that `_create()` is mostly used in netdata.

So the function name (which was suggested  by @ktsaou) seems correct to me.

simonnagl replied:
You are right. Did not get you `heartbeat_t` should be opaque. Fixed everything. Thank you for the review.

rlefevre commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
+ *
+ * \todo Maybe we should change the name (because init should have a corresponding free) to reset or even switch to use a define.
+ *
+ * @param hb Heartbeat to reset.
```
maybe `to reset` -> `to initialize or reset` ?

rlefevre commented on src/clocks.h:
```diff
@@ -1,108 +1,272 @@
+/**
+ * Sleeps until next multiple of `tick` using monotonic clock.
+ *
+ * After the call `hb` holds the current monotonic clock.
```
I would say this is an implementation detail, not exposed in the heartbeat API.

simonnagl commented:
Now every header file at least has TODO comments.
Also I made the commit history more clean.

Next steps:
- [ ] I will review this merge request myself and mark @ktsaou where I still need help to understand things.
- [ ] He will help me to finish this.
- [ ] Typo fixes and use of more consistent descriptions.

simonnagl commented on src/web_buffer.h:
```diff
@@ -1,77 +1,266 @@
+ *
+ * Use buffer_need_bytes() instead. It is quicker.
+ *
+ * \todo Do not extern this.
```
@ktsaou Do you agree?

ktsaou replied:
I don't follow. Do not extern `buffer_free()`? Why?

ktsaou replied:
I started a review by mistake. I'll try to cancel it...

simonnagl replied:
> I don't follow. Do not extern buffer_free()? Why?

Because buffer_need_bytes() provides the same functionality but is quicker. Tell me if I am wrong.

simonnagl commented on src/web_buffer.h:
```diff
@@ -1,77 +1,266 @@
 // options
-#define WB_CONTENT_CACHEABLE            1
-#define WB_CONTENT_NO_CACHEABLE         2
+#define WB_CONTENT_CACHEABLE            1 ///< ktsaou: Your help needed
```
@ktsaou Your help needed.

ktsaou replied:
`WB_CONTENT_CACHEABLE` instructs the web server to respond with HTTP headers stating that the response can be cached by proxy servers and web clients for some time (currently fixed to 1 day).

simonnagl commented on src/web_buffer.h:
```diff
@@ -1,77 +1,266 @@
-#define WB_CONTENT_CACHEABLE            1
-#define WB_CONTENT_NO_CACHEABLE         2
+#define WB_CONTENT_CACHEABLE            1 ///< ktsaou: Your help needed
+#define WB_CONTENT_NO_CACHEABLE         2 ///< ktsaou: Your help needed
```
@ktsaou Your help needed.

ktsaou replied:
`WB_CONTENT_NO_CACHEABLE` instructs the web server to respond with HTTP headers stating that the response should never be cached, not even for a single second. This is used for API responses.

simonnagl commented on src/web_buffer.h:
```diff
@@ -1,77 +1,266 @@
+#define CT_APPLICATION_JSON             1  ///< MIME Type application/json
+#define CT_TEXT_PLAIN                   2  ///< MIME Type text/plain
+#define CT_TEXT_HTML                    3  ///< MIME Type text/html
+#define CT_APPLICATION_X_JAVASCRIPT     4  ///< MIME Type application/javascript. \todo We should switch this to CT_APPLICATION_JAVASCRIPT
```
I did some research which are the correct types. I have no motivation to fix this now. If you think this may be fixed I leave the todos in. Check the other types too...

ktsaou replied:
What is wrong?

ktsaou replied:
I think we should read/provide an file similar to `mime.types`.

simonnagl replied:
Can you please make clear which file you mean?.
Nothing is really wrong with it. But we might use names matching to the official types:
https://www.iana.org/assignments/media-types/media-types.xhtml

simonnagl commented on src/web_buffer_svg.h:
```diff
@@ -1,7 +1,42 @@
+ *
+ * If unit is one of `ok/error`, `ok/failed`, `up/down`, `on/off` we interpret value as boolean and print the representating name. I.e. `ok`
+ *
+ * \todo ktsaou: Your help needed - document precision
```
@ktsaou Your help needed

ktsaou replied:
`precision` is the number of fractional digits the value should have. A negative number means auto, that is:

- ` > 1000` -> 0
- ` > 10` -> 1
- ` > 0.1` -> 2
- else 4

simonnagl commented on src/plugin_proc.h:
```diff
@@ -1,31 +1,232 @@
+ * @brief Thread to collect linux proc filesystem metrics.
+ *
+ * __How to add a data collector__
+ * \todo Describe how to add a data collector.
```
I will leave this todo there if you are OK with it. This is a time consuming work and I do not see the need to have it in this pull request.

simonnagl commented on src/registry_person.h:
```diff
@@ -1,60 +1,164 @@
+ * Reallocate REGISTRY_PRSON_URL.
+ *
+ * Needed to change the name of a PERSON_URL.
+ * \todo ktsaou: Your help needed. Do we want to expose this?
```
@ktsaou Your help needed

simonnagl replied:
ping

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
+/**
+ * Free RRDDIM of RRDSET.
+ *
+ * \todo ktsaou: Your help needed. Is it save to extern this or should this only be called from rrdset_free_all()?
```
@ktsaou Your help needed.

simonnagl replied:
ping

ktsaou commented on src/README.md:
```diff
@@ -0,0 +1,26 @@
+
+# Overview
+
+\todo This is like nothing. Extend or remove.
```
ok. postpone.

ktsaou commented on src/adaptive_resortable_list.h:
```diff
@@ -32,89 +31,143 @@
+ * \skip arl_callback_str2ull
+ * \until }
+ *
+ * \todo How to choose a value for rechecks?
```
Let's assume a file has 10 keywords. We are only interested for the first 3. ARL will stop parsing the file when all 3 are satisfied (i.e. it will not even read the 4th keyword).

After some time, the file has 9 keywords and one of the wanted ones is missing. ARL needs 3, it does not find them, so it parses the whole file again and again.

`rechecks` provides the number of iterations, after which ARL will re-examine the situation and decide what is best to do. It does not do this all the time, since it requires a complete walk through of the ARL keywords.

So, with `rechecks = 2`, ARL will do this complete walk through every 2 iterations and decide on the number of wanted keywords.

The same could happen for the opposite condition. The file has 10 keywords, we are interested for 4, but only 3 are in the file, then after some time the 4th keyword is added. ARL may not "see" the added keyword, for `recheck` number of iterations.

Too small `recheck` and CPU resources are wasted.
Too big `recheck` and information may be missing from the charts.

ktsaou commented on src/eval.h:
```diff
@@ -1,73 +1,114 @@
+ * @file eval.h
+ * @brief Expression handling.
+ *
+ * \todo How to use it?
```
https://github.com/firehol/netdata/wiki/health-configuration-reference#expressions

ktsaou commented on src/log.h:
```diff
@@ -1,73 +1,211 @@
 
-extern time_t error_log_throttle_period, error_log_throttle_period_backup;
-extern unsigned long error_log_errors_per_period;
+extern time_t error_log_throttle_period;          ///< ktsaou: Your help needed.
```
`error_log_throttle_period` gives the time in seconds the log should be suspended. Log throttling is not rolling. So, if `error_log_throttle_period = 3600`, the log throttling variables are reset once every hour.

`error_log_throttle_period_backup` is used to restore throttling in case netdata suspends this functionality and needs to resume it. Error throttling is suspended when saving the database (so, you will always see these errors).

`error_log_errors_per_period` this is the threshold. If that many errors are logged, error logging is suspended for the end of the period.

ktsaou commented:
Answered them all.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+#define RRDCALC_FLAG_CALC_ERROR            0x00000008 ///< Calulation error occured
+#define RRDCALC_FLAG_WARN_ERROR            0x00000010 ///< Error level warning occured
+#define RRDCALC_FLAG_CRIT_ERROR            0x00000020 ///< error level error occured
+#define RRDCALC_FLAG_RUNNABLE              0x00000040 ///< ktsaou: Your help needed.
```
@ktsaou: Your help needed

ktsaou replied:
`RRDCALC_FLAG_RUNNABLE` is used to cache the response of `rrdcalc_isrunnable()`. It is just an internal optimization for `health_main()`.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+#define RRDCALC_FLAG_WARN_ERROR            0x00000010 ///< Error level warning occured
+#define RRDCALC_FLAG_CRIT_ERROR            0x00000020 ///< error level error occured
+#define RRDCALC_FLAG_RUNNABLE              0x00000040 ///< ktsaou: Your help needed.
+#define RRDCALC_FLAG_NO_CLEAR_NOTIFICATION 0x80000000 ///< ktsaou: Your help needed.
```
@ktsaou: Your help needed

ktsaou replied:
There are a few alarms that should not send a CLEAR notification. For example, we try to find a web attack by comparing two 10 second durations: the first starting at `-10s` and the second starting at `-20s`.

If the first is 10x the second, we assume we are under attack. In 10 seconds however, data are shifted and although the attack continues, we see that now the first and the second are equal, so we send a CLEAR notification, although the attack continues.

So, `RRDCALC_FLAG_NO_CLEAR_NOTIFICATION` prevents netdata from sending CLEAR notifications.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
 } RRDCALC;
 
-#define RRDCALC_HAS_DB_LOOKUP(rc) ((rc)->after)
+#define RRDCALC_HAS_DB_LOOKUP(rc) ((rc)->after) ///< ktsaou: Your help needed
```
@ktsaou: Your help needed

ktsaou replied:
`RRDCALC_HAS_DB_LOOKUP` should state if the alarm has a `lookup` line or not.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+#define HEALTH_ENTRY_FLAG_EXEC_RUN              0x00000004 ///< Health entry prcessing successful.
+#define HEALTH_ENTRY_FLAG_EXEC_FAILED           0x00000008 ///< Health entry processing failed.
+#define HEALTH_ENTRY_FLAG_SAVED                 0x10000000 ///< Health entry saved. 
+#define HEALTH_ENTRY_FLAG_NO_CLEAR_NOTIFICATION 0x80000000 ///< ktsaou: Your help needed.
```
@kt

ktsaou replied:
same as above for the alarm log.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+ * @param st RRDSET to create RRDSETVAR for.
+ * @param variable name.
+ * @param type RRDVAR_TYPE_*
+ * @param value ktsaou: Your help needed.
```
@ktsaou Your help needed

ktsaou replied:
RRDVAR has a `void *value`. RRDVAR_TYPE_* state the type of data the pointer points to.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+ * @param variable name.
+ * @param type RRDVAR_TYPE_*
+ * @param value ktsaou: Your help needed.
+ * @param options ktsaou: Your help needed.
```
@ktsaou Your help needed

ktsaou replied:
hm... field `options` in `RRDSETVAR` seems unused.

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+ * @param type RRDVAR_TYPE_*
+ * @param prefix of variable name.
+ * @param suffix of variable name.
+ * @param value ktsaou: Your help needed.
```
@ktsaou Your help needed

ktsaou replied:
well... `value` is a pointer to the value of the variable...

simonnagl commented on src/health.h:
```diff
@@ -111,262 +129,405 @@ typedef struct rrddimvar {
+ * @param prefix of variable name.
+ * @param suffix of variable name.
+ * @param value ktsaou: Your help needed.
+ * @param options ktsaou: Your help needed.
```
@ktsaou Your help needed

ktsaou replied:
also unused

ktsaou commented:
@simonnagl RRDVARs (of all kinds), should be viewed as metadata around existing netdata variables. So, we just name them and we index them. We don't copy them, but instead we keep pointers to the real variables, inside the real netdata structures (charts and dimensions).

The result is that netdata works full speed while collecting data, using natively the variables in its structures (charts and dimensions), but we have several indexes with metadata, which we can use to lookup these variables. These indexes are used by alarm expressions. This is how alarms can use in `calc`, `warn` and `crit` lines all the variables.

ktsaou commented:
(I know it sounds complicated - I wrote it 2 times to get it right... :)

simonnagl commented on src/registry_internals.h:
```diff
@@ -3,90 +3,184 @@
+/**
+ * ktsaou: Your help needed.
+ *
+ * @return ktsaou: Your help needed.
```
I understand that log is something like a DB. Can you please clarify what's the difference between registry_log_load and registry_db_load?

ktsaou replied:
of course.

`registry_db` is a dump of the registry database in memory.
`registry_log` is the transaction log, that alters the database in memory.

So, for a variable `X` in the database that is updated 10 times, you will have 10 rows in `registry_log` immediately after the value is changed.

When 1.000.000 (configurable) rows are written to `registry_log`, the database is dumped again in `registry_db` and `registry_log` is rotated.

This way, even if netdata crashes, transactions are not lost. When netdata starts, it loads `registry_db` and then `registry_log`, to continue from the point it was.

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
 
-typedef long long total_number;
-#define TOTAL_NUMBER_FORMAT "%lld"
+typedef long long total_number;    ///< ktsaou: Your help needed.
```
@ktsaou Your help needed.

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
 extern RRDSET *rrdset_find_byname(const char *name);
 
+/**
+ * ktsaou: Your help needed.
```
@ktsaou Your help needed.

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
+/**
+ * ktsaou: Your help needed.
+ *
+ * @param st RRDSET.
```
@ktsaou Your help needed

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
+ */
 extern void rrdset_next_usec(RRDSET *st, usec_t microseconds);
+/**
+ * ktsaou: Your help needed.
```
@ktsaou Your help needed

simonnagl commented on src/rrd.h:
```diff
@@ -1,414 +1,720 @@
+ * @param st RRDSET of RRDDIM.
+ * @param rd RRDDIM.
+ * @param value to add.
+ * @return ktsaou: Your help needed. Last interpolated value?
```
@ktsaou: Your help needed

simonnagl commented on src/rrd2json.h:
```diff
@@ -1,78 +1,216 @@
+#define GROUP_MIN 2             ///< (a,b) => a>b?b:a
+#define GROUP_MAX 3             ///< (a,b) => a>b?a:b
+#define GROUP_SUM 4             ///< (a,b) => a+b
+#define GROUP_INCREMENTAL_SUM 5 ///< ktsaou: Your help needed.
```
@ktsaou: Your help needed.

simonnagl commented on src/unit_test.h:
```diff
@@ -1,8 +1,30 @@
+/**
+ * Run all unit tests.
+ *
+ * @param delay ktsaou: Your help needed
```
@ktsaou: Your help needed

simonnagl commented on src/unit_test.h:
```diff
@@ -1,8 +1,30 @@
+ * Run all unit tests.
+ *
+ * @param delay ktsaou: Your help needed
+ * @param shift kstaou: Your help needed
```
@ktsaou: Your help

ktsaou commented:
@simonnagl I know you wait for me. I need to focus on multi-host support (which unfortunately will break a lot of your changes - sorry...). Give me a few days to finish it.

simonnagl commented:
No problem!

simonnagl commented:
I see many many changes. I'll start rebasing now. I do not know if I can finish this soon.
When I adapted the changes to current master I will notify you.

If you agree even if there are some points left I would like you to merge this. There won't be wrong documentation but missing one. I will change to TODOs instead of ktsaou: Your help needed. These can be done in feature merge requests then.

We can discuss this if it is mergeable again.

paulfantom commented:
@cakrit look at this.

simonnagl commented:
@cakrit If you are interested in this I can adopt this PR for the new directory structure. Just tell me.

cakrit commented:
Thanks @simonnagl. Although I won't be able to look into this in the short term, you're probably right and it won't be possible to proceed at all if it's not adapted to the new structure. I'd be a bit hesitant to take up your time before we've decided to proceed with this though. We'll have a new dev joining in Jan, so I'll discuss it with him first.

simonnagl commented:
I will close this for now. If you want to have this in Netdata please contact me. I am willing to provide a setup with the basic work again.
